Tests/ContentKeyPeriodCrudTests.cs

```diff
@@ -0,0 +1,228 @@
+		{
+			var ex = Assert.Throws<XmlSchemaValidationException>(() =>
+				CpixDocument.Load(new MemoryStream(Encoding.UTF8.GetBytes("<?xml version=\"1.0\" encoding=\"utf-8\"?><CPIX xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\" xmlns=\"urn:dashif:org:cpix\" xmlns:ds=\"http://www.w3.org/2000/09/xmldsig#\" xmlns:enc=\"http://www.w3.org/2001/04/xmlenc#\" xmlns:pskc=\"urn:ietf:params:xml:ns:keyprov:pskc\"><ContentKeyPeriodList><ContentKeyPeriod id=\"1cannotstartwithnumber\" index=\"1\" /></ContentKeyPeriodList></CPIX>"))));
+
+			Assert.Contains("invalid according to its datatype", ex.Message);
+		}
+
+		[Fact]
+		public void Save_WithCpixContainingContentKeyPeriodsWithNonUniqueIds_FailsSchemaValidation()
+		{
+			var document = new CpixDocument();
+			document.ContentKeyPeriods.Add(new ContentKeyPeriod { Id = "duplicate", Index = 1 });
+			document.ContentKeyPeriods.Add(new ContentKeyPeriod { Id = "duplicate", Index = 2 });
+
+			var ex = Assert.Throws<XmlSchemaValidationException>(() => document.Save(new MemoryStream()));
+			Assert.Contains("is already used as an ID", ex.Message);
+		}
+
+		[Fact]
+		public void Load_WithCpixContainingContentKeyPeriodsWithNonUniqueIds_FailsSchemaValidation()
+		{
+			var ex = Assert.Throws<XmlSchemaValidationException>(() =>
+				CpixDocument.Load(new MemoryStream(Encoding.UTF8.GetBytes("<?xml version=\"1.0\" encoding=\"utf-8\"?><CPIX xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\" xmlns=\"urn:dashif:org:cpix\" xmlns:ds=\"http://www.w3.org/2000/09/xmldsig#\" xmlns:enc=\"http://www.w3.org/2001/04/xmlenc#\" xmlns:pskc=\"urn:ietf:params:xml:ns:keyprov:pskc\"><ContentKeyPeriodList><ContentKeyPeriod id=\"id1\" index=\"1\" /><ContentKeyPeriod id=\"id1\" index=\"2\" /></ContentKeyPeriodList></CPIX>"))));
+
+			Assert.Contains("is already used as an ID", ex.Message);
+		}
+		
+		[Fact]
+		public void Roundtrip_WithSignedCollectionOfVariousValidContentKeyPeriods_Succeeds()
+		{
+			var document = new CpixDocument();
+
+			var expectedPeriod1Index = 1;
+			var expectedPeriod2Id = "period_2";
+			var expectedPeriod2Index = 2;
+			var expectedPeriod3Id = "period_3";
+			var expectedPeriod3Start = DateTime.Now;
+			var expectedPeriod3End = DateTime.Now.AddHours(1);
+
+			document.ContentKeyPeriods.Add(new ContentKeyPeriod { Index = expectedPeriod1Index });
+			document.ContentKeyPeriods.Add(new ContentKeyPeriod { Id = expectedPeriod2Id, Index = expectedPeriod2Index});
+			document.ContentKeyPeriods.Add(new ContentKeyPeriod { Id = expectedPeriod3Id, Start = expectedPeriod3Start, End = expectedPeriod3End });
+			document.ContentKeyPeriods.AddSignature(TestHelpers.Certificate1WithPrivateKey);
+
+			document = TestHelpers.Reload(document);
+
+			Assert.Equal(3, document.ContentKeyPeriods.Count);
+
+			Assert.Null(document.ContentKeyPeriods.ElementAt(0).Id);
```

Comment:
Usually, we add a massage. Should it apply here? Or it's something we use internally.

Reply:
Good idea, but since none of the existing tests use messages either, I'm gonna keep as is.